File: src/main/java/com/github/zxh/classpy/classfile/bytecode/Instruction.java
package com.github.zxh.classpy.classfile.bytecode;

import com.github.zxh.classpy.classfile.ClassComponent;
import com.github.zxh.classpy.classfile.ClassFileReader;
import com.github.zxh.classpy.classfile.jvm.Opcode;

/**
 * Base class for all instructions.
 */
public class Instruction extends ClassComponent {

    protected final Opcode opcode;
    protected final int pc;

    public Instruction(Opcode opcode, int pc) {
        this.opcode = opcode;
        this.pc = pc;
        setDesc(opcode.name());
    }

    public int getPc() {
        return pc;
    }
    
    @Override
    protected final void readContent(ClassFileReader reader) {
        if (!super.getComponents().isEmpty()) {
            super.readContent(reader);
        } else {
            reader.readUnsignedByte(); // opcode
            readOperands(reader);
        }
    }
    
    protected void readOperands(ClassFileReader reader) {
        if (opcode.operandCount > 0) {
            reader.skipBytes(opcode.operandCount);
        }
    }

}

File: src/main/java/com/github/zxh/classpy/classfile/bytecode/Instruction.kt
package com.github.zxh.classpy.classfile.bytecode

import com.github.zxh.classpy.classfile.ClassComponent
import com.github.zxh.classpy.classfile.constant.ConstantPool
import com.github.zxh.classpy.classfile.reader.ClassReader

/**
 * Base class for all instructions.
 */
open class Instruction(val opcode: Opcode, val pc: Int) : ClassComponent() {

    override fun readContent(reader: ClassReader) {
        if (super.subComponents == null || super.subComponents!!.isEmpty()) {
            super.readContent(reader);
        } else {
            reader.readUnsignedByte(); // opcode
            readOperands(reader);
        }
    }
    
    protected open fun readOperands(reader: ClassReader) {
        if (opcode.operandCount > 0) {
            reader.skipBytes(opcode.operandCount);
        }
    }

    override fun afterRead(cp: ConstantPool) {
        //val subComponents = super.subComponents;
        if (subComponents?.size == 2) {
            val operand = subComponents!![1]
            desc = opcode.name + operand.desc;
        } else {
            // todo
            desc = opcode.name
        }
    }

}
